hatchways/team-dumplings
# Implementation / integration  of the `Profile payment`/`checkout` #22

### What this PR :
##  Summary
- To satisfied, the requirement of the payment `customerId` attribute added to the `profile` model, 
- When a user navigates to the payment page for the first time, a stripe customer will be created for the authenticated user if `customerId` is not defined yet, 
- Each payment profile (payment method) created will be attached to the stripe customer as a source, 
- On the payment page, the frontend will fetch the payment methods from the stripe server 
- Checkout creates a Payment intent, that will be confirmed on the frontend using a client secret, after successful payment confirmation, the request status will be updated to `paid`
  
- Functionalities:  
  - Payment profiles UI
    - Payment profiles (payment methods) integration 
    - Add a new `payment method`
    - Set default `payment method`
  - Checkout UI
    - Checkout integration with the `booking` page
    - Popup `successful`, `error` message in the `booking` page, after payment confirmation

### Screenshots / Videos :

>Select payment options
>![select payment options](https://user-images.githubusercontent.com/32603079/139597508-bd7b3184-c511-4bbe-8b5d-79e2d5660377.gif)

>Add new payment method
>![add new payment method](https://user-images.githubusercontent.com/32603079/139597515-b85da392-c884-4695-865c-cc40fe0cd684.gif)

>Add second payment method
>![add second payment method](https://user-images.githubusercontent.com/32603079/139597527-c05a8682-0552-4daa-8634-0ff5e791635f.gif)

>Payment method added
>![payment method added](https://user-images.githubusercontent.com/32603079/139597540-3711b92e-2aab-4841-a2c8-7d89ef664b98.gif)

> Select default payment method
>![Select default payment method](https://user-images.githubusercontent.com/32603079/139597501-cfee2436-214f-44d8-8a6f-8ff0b730218c.gif)

> successful payment message
>![confirm payment](https://user-images.githubusercontent.com/32603079/139597532-ae65122a-3be4-4dec-bfa7-c3b77d4f5668.gif)


### Information needed to test this feature:
- Mock-data on the database for the `Requests` document
- Setup `private` , `public` stripe key
- Users roles should be set to 'sitter'
- Start the app and navigate 
  - to `/booking`, to test the `checkout`
  - to `/payment`, to test the `payment` profiles

### Limitations with the current functionality:
- Refund
- Payment cancelation
- Payout (exp: sitter)

### Features:
- [x] #34 
- [x] #22
- [ ] Payout 
- [ ] Reinforce security
